openshift-storage-libs/openshiftstoragelibs/baseclass.py

```diff
@@ -806,6 +924,138 @@ def match_volume_by_prefix(self, prefix):
+
+        # Fetch heketi volume list before operation
+        h_vols_initial = heketi_volume_list(
+            self.heketi_client_node, self.heketi_server_url,
+            json=True).get('volumes')
+
+        # Add cleanup function to clean stale volumes created during test
+        self.addCleanup(self.cleanup_heketi_volumes, h_vols_initial)
+
+        # Fetch heketi blockvolume list before operation
+        h_vols_count_initial = len(heketi_blockvolume_list(
+            self.heketi_client_node, self.heketi_server_url,
+            json=True).get('blockvolumes'))
+
+        # Temporary replace g.run with g.async_run in heketi_blockvolume
+        # create func to be able to run it in background. Also, avoid parsing
+        # the output as it won't be json at that moment. Parse it after
+        # reading the async operation results.
+        def run_async(cmd, hostname, raise_on_error=True):
+            async_op = g.run_async(host=hostname, command=cmd)
+            async_obj.append(async_op)
+            return async_op
+
+        # Waiter for overall volumes
+        _waiter = Waiter(timeout=timeout, interval=interval)
+        while vol_count > 0:
+            # Change batch_amount if pvc_count < batch_amount
+            if vol_count < batch_amount:
+                batch_amount = vol_count
+            for _ in range(batch_amount):
+                with mock.patch.object(
+                        json, 'loads', side_effect=(lambda j: j)):
+                    with mock.patch.object(
+                            command, 'cmd_run', side_effect=run_async):
+                        heketi_blockvolume_create(
+                            self.heketi_client_node, self.heketi_server_url,
+                            size, json=True, **kwargs)
+            h_vols_count_initial += batch_amount
+
+            # Wait for all the blockvolume to be created
+            for w in _waiter:
+                h_vols_count_final = len(heketi_blockvolume_list(
+                    self.heketi_client_node, self.heketi_server_url,
+                    json=True).get('blockvolumes'))
+                if h_vols_count_final == h_vols_count_initial:
+                    break
+            if w.expired:
+                # Delete all the blockvolume i.e created before failure
+                h_vols = heketi_blockvolume_list(
+                    self.heketi_client_node, self.heketi_server_url,
+                    json=True).get('blockvolumes')
+                for vol in h_vols:
+                    self.addCleanup(
+                        heketi_blockvolume_delete,
+                        self.heketi_client_node, self.heketi_server_url, vol)
+
+                raise AssertionError(
+                    "Failed to create blockvolumes after waiting for {} "
+                    "secs".format(timeout))
+            vol_count -= batch_amount
+
+        # Check that all background processes got exited
+        for obj in async_obj:
```

Comment:
Move this for loop inside while loop  where volumes get created successfully.

Reply:
done